### PR TITLE
Rrb tree pseq

### PR DESCRIPTION
**Pull request by macrologist**
@stylewarning @eliaslfox @colescott  

This PR adds a Relaxed Radix Balanced Tree implementation of persistent sequences to the coalton standard library. 

The rebalancing step attempts to minimize memory allocation, reusing already allocated contiguous memory whenever possible.

### TODO

- [x] Need to fix a bug in the radix search (change +1 to linear search) 
- [x] Test that branching factor invariants are preserved in various cases 
- [x] More typeclass impls (e.g. Eq)

### CODE REVIEW

**Comment by stylewarning** [comment]
Note: This would supersede my Seq PR.

**Review by eliaslfox — APPROVED** [review]
This looks really good.

The package should be added to [this list](https://github.com/coalton-lang/coalton/blob/main/src/doc/generate-documentation.lisp#L101-L135) so that it shows up in the generated documentation.

Approving but waiting on @colescott and @stylewarning's reviews before merging.

File: library/pseq.lisp
Comment (macrologist) [reply library/pseq.lisp:223]: I can't seem to do it.  

`RuntimeRepr :a` evidently needs to be satisfied for `PSeq :a` but I don't know how to say that the constraint needs to be satisfied for the arguments of the type constructor `PSeq`

Comment (macrologist) [reply library/pseq.lisp:223]: (I imagine that this is part of the reason that `Vector` isn't a `Monad`?)

Comment (eliaslfox) [reply library/pseq.lisp:227]: Oh yeah, nvm then.

**Comment by stylewarning** [comment]
Bike Shed (TM): Could I suggest just the name "Seq"? Matches Clojure, also "p" can be interpreted as "parallel" (cf. plet pdotimes etc).

**Comment by stylewarning** [comment]
Bike Shed (TM): Could I suggest just the name "Seq"? Matches Clojure, also "p" can be interpreted as "parallel" (cf. plet pdotimes etc).

**Comment by macrologist** [comment]
if @colescott and @stylewarning have any (more) complaints, lemme know. Otherwise @eliaslfox (or whomever) can merge.

File: library/seq.lisp
Comment (stylewarning) [inline library/seq.lisp:514]: I would avoid consing the list. Maybe just expand into a chain of PUSH, to match how MAKE-LIST works?

File: library/seq.lisp
Comment (stylewarning) [inline library/seq.lisp:53]: I wish we had a way to define constants in Coalton.

File: library/seq.lisp
Comment (stylewarning) [inline library/seq.lisp:83]: Coalton docstring style is to use markdown syntax and lowercase names

    Get the member of `seq` at ...

File: library/seq.lisp
Comment (stylewarning) [inline library/seq.lisp:69]: If we get rid of the runtimerepr stuff, we should make this LOAD-TIME-VALUEd

**Review by stylewarning — APPROVED** [review]
My only grievance is the impl of MAKE. Eventually it would be nice to "statically" optimize MAKE and other conversion functions to allocate the tree structure wholesale without intermediate consing—but not for this PR.

I'm fine if we punt changing the definition of MAKE in the interest of moving this forward, and having follow up PRs to improve things. m

File: library/seq.lisp
Comment (macrologist) [reply library/seq.lisp:69]: We'd also be able to define `Monad` for `Vector` and consequently `Seq`.

**Comment by macrologist** [comment]
I Tried writing an optimized macro - but it still uses `conc` to make it somewhat sane. However, the `runtimerepr` constraint make sit not typecheck on high-arity invocations of `make`.  Still working on it.

**Comment by macrologist** [comment]
@stylewarning Ok I added an experimental extra commit that changes `seq:make` to build sequences at "compile time". 

The reason it creates such a massive quantity of `coalton:let` bindings is because I was not sure that I could build vectors with coalton expressions inside of them and have it do the right thing. 

That is, I want to be able to write `(seq:make (+ 1 1) (+ 2 2))`.

If there's a better way than the approach I took, I'm all ears. The final case in the macro is the "just freaking concatenate" case.

If nobody objects to the horror of this macro, then I'll squash the commits that add it into the branch to mrege, otherwise I'll abandon them.

**Pull request merged**
